pallets/ajuna-tournament/src/lib.rs

```diff
@@ -0,0 +1,753 @@
+				TournamentState::ActivePeriod(tournament_id) |
+				TournamentState::ClaimPeriod(tournament_id, _) =>
+					Tournaments::<T, I>::get(season_id, tournament_id),
+			}
+		}
+
+		fn is_golden_duck_enabled_for(season_id: &T::SeasonId) -> bool {
+			match ActiveTournaments::<T, I>::get(season_id) {
+				TournamentState::Inactive => false,
+				TournamentState::ActivePeriod(tournament_id) |
+				TournamentState::ClaimPeriod(tournament_id, _) => {
+					matches!(
+						GoldenDucks::<T, I>::get(season_id, tournament_id),
+						GoldenDuckStateFor::<T, I>::Enabled(_, _)
+					)
+				},
+			}
+		}
+
+		fn get_treasury_account_for(season_id: &T::SeasonId) -> AccountIdFor<T> {
+			Self::tournament_treasury_account_id(*season_id)
+		}
+	}
+
+	impl<T: Config<I>, I: 'static>
+		TournamentMutator<AccountIdFor<T>, T::SeasonId, BlockNumberFor<T>, BalanceOf<T, I>>
+		for Pallet<T, I>
+	{
+		fn try_create_new_tournament_for(
+			creator: &AccountIdFor<T>,
+			season_id: &T::SeasonId,
+			config: TournamentConfigFor<T, I>,
+		) -> Result<TournamentId, DispatchError> {
+			Self::ensure_valid_tournament(season_id, &config)?;
+
+			let next_tournament_id =
+				NextTournamentIds::<T, I>::mutate(season_id, |tournament_id| {
+					let assigned_id = *tournament_id;
+					*tournament_id = tournament_id.saturating_add(1);
+					assigned_id
+				});
+
+			Self::try_insert_tournament_schedule(season_id, &next_tournament_id, &config)?;
+
+			if let Some(reward) = config.initial_reward {
+				let treasury_account = Self::tournament_treasury_account_id(*season_id);
+				T::Currency::transfer(
+					creator,
+					&treasury_account,
+					reward,
+					ExistenceRequirement::KeepAlive,
+				)?;
+			}
+
+			if let GoldenDuckConfig::Enabled(percentage) = config.golden_duck_config {
+				GoldenDucks::<T, I>::insert(
+					season_id,
+					next_tournament_id,
+					GoldenDuckStateFor::<T, I>::Enabled(percentage, None),
+				);
+			}
+
+			Tournaments::<T, I>::insert(season_id, next_tournament_id, config);
+
+			Self::deposit_event(Event::<T, I>::TournamentCreated {
+				season_id: *season_id,
+				tournament_id: next_tournament_id,
+			});
+
+			Ok(next_tournament_id)
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> TournamentRanker<T::SeasonId, T::RankedEntity, T::EntityId>
+		for Pallet<T, I>
+	{
+		fn try_rank_entity_in_tournament_for<R>(
+			season_id: &T::SeasonId,
+			entity_id: &T::EntityId,
+			entity: &T::RankedEntity,
+			ranker: &R,
+		) -> DispatchResult
+		where
+			R: EntityRank<EntityId = T::EntityId, Entity = T::RankedEntity>,
+		{
+			let tournament_id = Self::try_get_active_tournament_id_for(season_id)?;
+			let tournament_config = Self::get_active_tournament_config_for(season_id)
+				.ok_or::<Error<T, I>>(Error::<T, I>::TournamentNotFound)?;
+
+			TournamentRankings::<T, I>::mutate(season_id, tournament_id, |table| {
+				if let Err(index) = table.binary_search_by(|(other_id, other)| {
+					ranker.rank_against((entity_id, entity), (other_id, other))
+				}) {
```

Comment:
Maybe the whole thing is a bit better readable if the show the ok statement before:

```rust
match table.binary_search_by(|(other_id, other)| ranker.rank_against((entity_id, entity), (other_id, other))) {
   Ok(_) => Ok(()),
   Err(index) => {....}
}
```